hbase-migration-tools/bigtable-hbase-replication/bigtable-hbase-1.x-replication/src/test/java/com/google/cloud/bigtable/hbase1_x/replication/HbaseToCloudBigtableBidirectionalReplicationEndpointTest.java

```diff
@@ -0,0 +1,197 @@
+
+import static com.google.cloud.bigtable.hbase.replication.utils.TestUtils.CF2;
+
+import com.google.cloud.bigtable.hbase.BigtableConfiguration;
+import com.google.cloud.bigtable.hbase.replication.utils.TestUtils;
+import com.google.cloud.bigtable.test.helper.BigtableEmulatorRule;
+import java.io.IOException;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.HBaseTestingUtility;
+import org.apache.hadoop.hbase.HColumnDescriptor;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.hbase.HTableDescriptor;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.Connection;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.client.RowMutations;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.client.replication.ReplicationAdmin;
+import org.apache.hadoop.hbase.replication.ReplicationEndpoint.ReplicateContext;
+import org.apache.hadoop.hbase.replication.ReplicationPeerConfig;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Test bidirectional replication.
+ * This test is separate from the other endpoint tests because it requires spinning up a
+ * cluster with additional config settings.
+ */
+@RunWith(JUnit4.class)
+public class HbaseToCloudBigtableBidirectionalReplicationEndpointTest {
+
+  public static class TestReplicationEndpoint extends HbaseToCloudBigtableReplicationEndpoint {
+
+    static AtomicInteger replicatedEntries = new AtomicInteger();
+
+    @Override
+    public boolean replicate(ReplicateContext replicateContext) {
+      boolean result = super.replicate(replicateContext);
+      replicatedEntries.getAndAdd(replicateContext.getEntries().size());
+      return result;
+    }
+  }
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(HbaseToCloudBigtableBidirectionalReplicationEndpointTest.class);
+
+  private static HBaseTestingUtility hbaseTestingUtil;
+  private static ReplicationAdmin replicationAdmin;
+
+  @ClassRule
+  public static final BigtableEmulatorRule bigtableEmulator = new BigtableEmulatorRule();
+
+  private static Connection cbtConnection;
+  private static Connection hbaseConnection;
+
+  private Table hbaseTable;
+  private Table cbtTable;
+
+  private static byte[] cbtQualifier = "customCbtQualifier".getBytes();
+  private static byte[] hbaseQualifier = "customHbaseQualifier".getBytes();
+  public static final byte[] ROW_KEY_2 = "test-row-2".getBytes();
+
+  @BeforeClass
+  public static void setUpCluster() throws Exception {
```

Comment:
These tests are slow. Can you reuse the existing tests and avoid the cost of setting up a new mini cluster?

Reply:
We need to enable bidirectional_replication_mode on HbaseTestingUtility constructor and before cluster start.

Reply:
Interesting. Is there any way to update the config on the fly? There should be a way. can you research a bit?

Reply:
@igorbernstein2 do you know if we can change the HBase config after initiating the mini-cluster?

Reply:
Closing this thread.